improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,392 @@
+            pressure = np.ma.masked_where(pressure == -9999, pressure)
+
+        pressure = pressure.reshape(r, l, m)
+        pressure_template.data = pressure
+
+        temperature = self.variable_at_pressure(temperature_on_pressure, pressure)
+
+        temperature = np.ma.masked_where(np.ma.getmask(pressure), temperature)
+        temperature_template.data = temperature
+        return pressure_template, temperature_template
+
+    def relative_humidity_at_268(
+        self, relative_humidity: Cube, pressure_at_268: Cube
+    ) -> Cube:
+        """Extract relative humidity at pressure of the environment at 268.15K"""
+
+        relative_humidity_data = self.variable_at_pressure(
+            relative_humidity, pressure_at_268.data
+        )
+        relative_humidity = pressure_at_268.copy(data=relative_humidity_data)
+        relative_humidity.rename("relative_humidity_at_268.15K")
+        relative_humidity.units = "kg/kg"
+
+        return relative_humidity
+
+    def temperature_after_saturated_ascent_from_CCL(
+        self,
+        ccl_temperature: Cube,
+        ccl_pressure: Cube,
+        pressure_at_268: Cube,
+        humidity_mixing_ratio_at_268: Cube,
+    ) -> np.ndarray:
+        """Calculates the temperature after a saturated ascent
+        from the cloud condensation level to the pressure of the atmosphere at 268.15K"""
+
+        t_dry = dry_adiabatic_temperature(
+            ccl_temperature.data, ccl_pressure.data, pressure_at_268.data
+        )
+        t_2, _ = adjust_for_latent_heat(
+            t_dry, humidity_mixing_ratio_at_268.data, pressure_at_268.data
+        )
+        return t_2
+
+    def dry_adiabatic_drop_to_ccl(
+        self, ccl_pressure: Cube, temperature_at_268: Cube, pressure_at_268: Cube
+    ) -> np.ndarray:
+        """Calculates a dry adiabatic drop from the pressure of the environment at
+        268.15K to the cloud condensation level."""
+
+        t_dry = dry_adiabatic_temperature(
+            temperature_at_268.data, pressure_at_268.data, ccl_pressure.data
+        )
+        return t_dry
+
+    def get_hail_size(self, vertical: np.ndarray, horizontal: np.ndarray):
+        """Uses the lookup_table and the vertical and horizontal indexes calculated
+        to extract and store values from the lookup nomogram. Masked data points or
+        if vertical or horizontal values are negative lead to a hail_size of 0."""
+
+        lookup_table = self.nomogram_values()
+        shape = np.shape(vertical)
+
+        horizontal_rounded = np.around(horizontal / 5, decimals=0) - 1
+        vertical_rounded = np.around(vertical * 2, decimals=0) - 1
+
+        horizontal_flat = horizontal_rounded.flatten(order="C")
+        vertical_flat = vertical_rounded.flatten(order="C")
+        hail_size_list = []
+
+        vertical_clipped = np.clip(vertical_flat, None, len(lookup_table) - 1)
+        horizontal_clipped = np.clip(horizontal_flat, None, len(lookup_table[0]) - 1)
```

Comment:
The above lines are pretty hard to follow if one does not know what's going on. Some extra comments would be useful.

Reply:
I've tried to add in some extra lines explaining it so it's hopfully atleast a bit clearer.